File: year-2023/script-language/assignment-08/ex_01.pl
#!/usr/bin/perl
# Hlib-Oleksandr Suliz, Script Language, group no.2


use strict;
use warnings;
use Getopt::Std;
use Cwd qw(abs_path);
use File::Basename qw(dirname);

require './number_checker.pm';

my %opts;
getopts('diehvLm', \%opts);

sub usage {
    print "Usage: $0 [-d|-i] [-e] file1 [file2 ...]\n";
    print "  -d    Count all numbers, including exponential formats\n";
    print "  -i    Count only integers\n";
    print "  -e    Exclude lines starting with # from processing\n";
}

if ($opts{h}) {
    usage();
    exit 0;
}

my $number_checker = NumberChecker->new();

my $number_option = '';
if ($opts{d}) {
    $number_option = '-d';
} elsif ($opts{i}) {
    $number_option = '-i';
}

my $exclude_comments = $opts{e} ? 1 : 0;

sub count_numbers {
    my ($file) = @_;
    open my $fh, '<', $file or die "Cannot open file '$file': $!";
    my $lines = 0;
    my $words = 0;
    my $chars = 0;
    my $integers = 0;
    my $all_numbers = 0;

    while (<$fh>) {
        chomp;
        next if ($exclude_comments && /^\s*#/);
        my @words_in_line = split /\s+/, $_;

        $lines++;
        $chars += length($_);

        foreach my $word (@words_in_line) {
            $words++;
            if ($number_option eq '-i' && $number_checker->is_integer($word)) {
                $integers++;
                $all_numbers++;
            } elsif ($number_option eq '-d' && $number_checker->is_real_number($word)) {
                $all_numbers++;
            } elsif ($number_option eq '' && ($number_checker->is_integer($word) || $number_checker->is_real_number($word))) {
                $all_numbers++;
            }
        }
    }

    close $fh;

    return {
        lines     => $lines,
        words     => $words,
        chars     => $chars,
        integers  => $integers,
        all_numbers => $all_numbers
    };
}

my $total_lines = 0;
my $total_words = 0;
my $total_chars = 0;
my $total_integers = 0;
my $total_all_numbers = 0;

foreach my $file (@ARGV) {
    if (-e $file) {
        my $stats = count_numbers($file);

        $total_lines += $stats->{lines};
        $total_words += $stats->{words};
        $total_chars += $stats->{chars};
        $total_integers += $stats->{integers};
        $total_all_numbers += $stats->{all_numbers};

        print "\nFile: $file -- lines: $stats->{lines}, words: $stats->{words}, characters: $stats->{chars}";
        print ", integers: $stats->{integers}" if $opts{i};
        print ", all numbers: $stats->{all_numbers}" if $opts{d};
        print "\n";
    } else {
        print "File '$file' does not exist.\n";
    }
}

if (@ARGV > 1) {
    print "\nALL FILES -- lines: $total_lines, words: $total_words, characters: $total_chars";
    print ", integers: $total_integers" if $opts{i};
    print ", all numbers: $total_all_numbers" if $opts{d};
    print "\n\n";
} else {
    print "\n";
}

File: year-2023/script-language/assignment-11/ex_01.py
#!/usr/bin/python3
# Hlib-Oleksandr Suliz, Script Language, group no.2


import os
import re
import sys

f = open(os.devnull, 'w')
sys.stdout = f


def process_file(file_path, options):
    counts = {
        'word': 0,
        'line': 0,
        'char': 0,
        'integer': 0,
        'all_numbers': 0
    }

    with open(file_path, 'r') as my_file:
        for line in my_file:
            if options['e'] and line.strip().startswith("#"):
                continue

            counts['line'] += 1
            counts['char'] += len(line)

            for word in line.split():
                if word:
                    counts['word'] += 1

                if word.isdigit() or re.match(r"^([+-]?)(?=\d|\.\d)\d*(\.\d*)?([EeQqDd^]([+-]?\d+))?$", word):
                    counts['integer'] += 1
                    counts['all_numbers'] += 1

    return counts


def print_file_statistics(file_path, counts, options):
    print("\nFile: ", file_path, " -- lines: ", counts['line'], ", words: ", counts['word'], ", characters: ",
          counts['char'], end=" ")

    if options['i']:
        print(", integers: ", counts['integer'], end=" ")

    if options['d']:
        print(", all numbers: ", counts['all_numbers'], end=" ")

    print()


def main():
    options = {'d': False, 'i': False, 'e': False}
    files_nr = 0
    w_counts = {
        'word': 0,
        'line': 0,
        'char': 0,
        'integer': 0,
        'all_numbers': 0
    }

    for arg in sys.argv[1:]:
        if arg.startswith('-'):
            if arg == '-d':
                options['d'] = True
            elif arg == '-i':
                options['i'] = True
            elif arg == '-e':
                options['e'] = True
        else:
            files_nr += 1

    for file_index in range(1, len(sys.argv)):
        if not sys.argv[file_index].startswith('-'):
            counts = process_file(sys.argv[file_index], options)

            for key in w_counts:
                w_counts[key] += counts[key]

            print_file_statistics(sys.argv[file_index], counts, options)

    if files_nr > 1:
        print("\nALL FILES -- lines: ", w_counts['line'], ", words: ", w_counts['word'], ", characters: ",
              w_counts['char'], end=" ")

        if options['i']:
            print(", integers: ", w_counts['integer'], end=" ")

        if options['d']:
            print(", all numbers: ", w_counts['all_numbers'], end=" ")

    print()


if __name__ == "__main__":
    main()
